### ISSUE TITLE
TimeZone Spoofing based on IP address

### ISSUE THREAD

**Issue by DGjone44** [kkapsner/CanvasBlocker#557]
Looking for feature to Change TimeZone/Spoof Time Zone based on IP address.
Current addon I am using is Chameleon from Serenblue. This extension query IP address from online database to convert IP to TimeZone.

![image](https://user-images.githubusercontent.com/17850550/120907932-7abc8280-c62b-11eb-8374-ddd10022da9f.png)

![image](https://user-images.githubusercontent.com/17850550/120907944-9031ac80-c62b-11eb-9a0c-be81fab24ec3.png)

![image](https://user-images.githubusercontent.com/17850550/120907988-e56dbe00-c62b-11eb-9ff0-a4bbbdd35cf3.png)

![image](https://user-images.githubusercontent.com/17850550/120908023-2796ff80-c62c-11eb-889a-328cb3deeaa1.png)

_Happy to donate $20 to your BTC address at main page._

**Comment by kkapsner** [comment]
Timezone faking is very complex (I'm not sure if it's even completely possible).
I just ran a very quick test with Chameleon and it managed to fake getTimezoneOffset but failed to fake the generated date string (@sereneblue):
![image](https://user-images.githubusercontent.com/1105157/121756946-398d0c80-cb1c-11eb-8b29-af32012aa39c.png)

So what is your exact use case? Do you use a VPN and want to match the timezone you are leaving it?

**Comment by DGjone44** [comment]
> 
> 
> Timezone faking is very complex (I'm not sure if it's even completely possible).
> I just ran a very quick test with Chameleon and it managed to fake getTimezoneOffset but failed to fake the generated date string (@sereneblue):
> ![image](https://user-images.githubusercontent.com/1105157/121756946-398d0c80-cb1c-11eb-8b29-af32012aa39c.png)
> 
> So what is your exact use case? Do you use a VPN and want to match the timezone you are leaving it?


Thanks for testing my supplied reference app. Yes I used vpn to test something and IP location changes. Matching timezone.

**Comment by DGjone44** [comment]
While checking with https://whoer.net/#extended it always shows System and Local time is matched. (While using @sereneblue /Chameleon app)

**Comment by sereneblue** [comment]
> Timezone faking is very complex (I'm not sure if it's even completely possible).

I agree with the above. It's not possible to really spoof this using webextensions. The best alternative is to launch Firefox with a modified timezone environment variable on Linux. (maybe possible on Windows?). What was wrong with the generated string?

**Comment by kkapsner** [comment]
@sereneblue: you're right. The text is consistent but I selected HST timezone... so the timezone spoofing is not working at all at the moment or did I do something wrong? I just selected a timezone in the browser action popup.

**Comment by sereneblue** [comment]
> @sereneblue: you're right. The text is consistent but I selected HST timezone... so the timezone spoofing is not working at all at the moment or did I do something wrong? I just selected a timezone in the browser action popup.

It looks like you're testing on 127.0.0.1? Chameleon is disabled for private IPs/localhost.

**Comment by kkapsner** [comment]
Ah... did not know that.

OK - then here is my point how chameleon does not protect your timezone:
```javascript
function twoDiggits(num){return num >= 10? num.toFixed(0): "0" + num.toFixed(0);}
var now = new Date();
var today = now.getFullYear() + "-" + twoDiggits(now.getMonth() + 1) + "-" + twoDiggits(now.getDate())
alert((new Date(today + " 00:00:00") - new Date(today)) / 60000);
```

... and then there are web workers...

PS: you're leaking that you modified the functions: `Function.prototype.toString.call(Date);`

**Comment by sereneblue** [comment]
> Ah... did not know that.
> 
> OK - then here is my point how chameleon does not protect your timezone:
> 
> ```js
> function twoDiggits(num){return num >= 10? num.toFixed(0): "0" + num.toFixed(0);}
> var now = new Date();
> var today = now.getFullYear() + "-" + twoDiggits(now.getMonth() + 1) + "-" + twoDiggits(now.getDate())
> alert((new Date(today + " 00:00:00") - new Date(today)) / 60000);
> ```
> 
> ... and then there are web workers...
> 
> PS: you're leaking that you modified the functions: `Function.prototype.toString.call(Date);`

Thanks for pointing those out. I'll look into the first issue but not sure if there's actually anything that can be done about it. The second issue should be resolved in a future update.

**Comment by Irvinehimself** [comment]
As suggested by [@sereneblue](https://github.com/sereneblue), according to the [Arch Firefox Wiki](https://wiki.archlinux.org/title/Firefox/Privacy#Change_browser_time_zone), you can set the time-zone for the browser to match the VPN output portal by setting the time-zone for the app like so : `TZ=UTC firefox`.

What I normally use,  either in a _.desktop_ file or from the terminal is: `TZ=<tz> firefox -P <profile> --new-instance`

Since I use _Network Manager_, `nmcli connection show --active | grep "vpn"` would give me the current VPN certificate. From that I could deduce the time-zone. In other words, it would be fairly trivial to write a small script to automatically launch a  _Firefox_ instance with a time-zone matching the current VPN portal. 

Although, having said that, since VPN's are now fairly ubiquitous, matching time-zone to VPN output portal seems like a lot of work for little gain. The whole idea is to hide amongst the masses. In other words, you want a time zone with the largest possible population matching your browser's other characteristics.

I have know idea how to do the above in Windows?
